java-checks/src/main/java/org/sonar/java/checks/ShiftOnIntOrLongCheck.java

```diff
@@ -0,0 +1,130 @@
+    } else {
+      AssignmentExpressionTreeImpl assignmentExpressionTree = (AssignmentExpressionTreeImpl) tree;
+      target = assignmentExpressionTree.variable();
+      shift = assignmentExpressionTree.expression();
+    }
+
+    String identifier = getIdentifierName(target);
+    boolean expectLong = treeTypeIsLong(target);
+    boolean expectInt = treeTypeIsInt(target);
+    boolean shiftIsMinus = shift.is(Kind.UNARY_MINUS);
+
+    if (shift.is(Kind.UNARY_MINUS, Kind.UNARY_PLUS)) {
+      shift = ((UnaryExpressionTree) shift).expression();
+    }
+
+    if (isShiftEvaluable(shift, expectInt || expectLong)) {
+      String value = ((LiteralTreeImpl) shift).value();
+      long numberBits = (shiftIsMinus ? -1 : 1) * (expectInt ? Integer.parseInt(value) : Long.parseLong(value));
+      long reducedNumberBits = numberBits % (expectLong ? 64 : 32);
+      insertIssue(tree, numberBits, reducedNumberBits, expectInt, identifier);
+    }
+  }
+
+  private void insertIssue(Tree tree, long numberBits, long reducedNumberBits, boolean expectInt, String identifier) {
+    if (reducedNumberBits == 0L) {
+      addIssue(tree, new StringBuilder("Remove this useless shift (multiple of ").append(expectInt ? "32" : "64").append(")").toString());
+    } else if (tooManyBits(numberBits, expectInt)) {
+      if (expectInt) {
+        String identifierMarker = identifier == null ? "use" : new StringBuilder("make \"").append(identifier).append("\"").toString();
+        addIssue(tree, new StringBuilder("Either ").append(identifierMarker).append(" a \"long\" or correct this shift to ").append(reducedNumberBits).toString());
+      } else {
+        addIssue(tree, new StringBuilder("Correct this shift to ").append(reducedNumberBits).toString());
+      }
+    }
+  }
+
+  private boolean tooManyBits(long numberBits, boolean isInt) {
+    long value = Math.abs(numberBits);
+    return (isInt && value >= 32) || (!isInt && value >= 64);
+  }
+
+  private boolean treeTypeIsInt(ExpressionTree tree) {
+    return treeTypeIs(tree, "int") || treeTypeIs(tree, "java.lang.Integer");
+  }
+
+  private boolean treeTypeIsLong(ExpressionTree tree) {
+    return treeTypeIs(tree, "long") || treeTypeIs(tree, "java.lang.Long");
+  }
+
+  private boolean treeTypeIs(ExpressionTree tree, String fullyQualifiedName) {
+    return (tree.is(Kind.IDENTIFIER) && ((IdentifierTreeImpl) tree).getSymbolType().is(fullyQualifiedName)) ||
+      (tree.is(Kind.ARRAY_ACCESS_EXPRESSION) && ((ArrayAccessExpressionTreeImpl) tree).getSymbolType().is(fullyQualifiedName)) ||
+      (tree.is(Kind.PARENTHESIZED_EXPRESSION) && ((ParenthesizedTreeImpl) tree).getSymbolType().is(fullyQualifiedName)) ||
+      (tree.is(Kind.METHOD_INVOCATION) && ((MethodInvocationTreeImpl) tree).getSymbolType().is(fullyQualifiedName));
+  }
+
+  private boolean isShiftEvaluable(ExpressionTree tree, boolean intOrLong) {
+    return intOrLong && tree.is(Kind.INT_LITERAL, Kind.LONG_LITERAL);
```

Comment:
if tree is an INT_LITERAL or a LONG_LITERAL then its type is necessarily int or long so  the boolean parameter is useless here.